### ISSUE TITLE
Density maps don't work on *some* computers

### ISSUE THREAD

**Issue by petebankhead** [qupath/qupath#776]
## Bug report

**Describe the bug**
When using QuPath v0.3.0-rc1 at the EMBO Advanced Methods in Bioimage Analysis course, density maps worked for most people - but not for all. Even a restart didn't help.

Curiously, setting the density radius to 0 shows *something*. This suggests the problems arise during image filtering, which uses the JavaCPP presets for OpenCV.

**To Reproduce**
I wish I knew. It might be hardware-dependent (or some other system conflict), but at this point I don't see the issue on any computer I have access to.

**Expected behavior**
Consistency across computers where QuPath runs.

**Desktop (please complete the following information):**
 - OS: Windows 8, Windows 10 (so far)
 - QuPath Version 0.3.0-rc1

**Additional context**
My guess is that it is may be related to OpenCV, since QuPath is using OpenCV to apply filters and thereby calculate local densities. However, classifiers still seem to work - so not everything is broken, adding to the mystery.

**Any more insights from anyone experiencing the bug would be very welcome!**

**Comment by Svidro** [comment]
Nothing in the log I guess. 
Applies to all Density Types and Interpolation?
Can objects still be created even if the Overlay doesn't show anything?

Any users that suffered from the problem that also have GitHub accounts and could respond if they find/test something?

**Comment by melvingelbard** [comment]
Ok, I've managed to reproduce this issue on one computer (out of 3 tested). The image used was `OS-1.ndpi`.

* OS: MacOs (Mojave)
* CPU: 2.3 GHz Intel Core i9
* 32 Gb RAM
* Radeon Pro 560X 4 Gb

Same problem as mentioned above. No overlay is shown, except when `density radius < 10`, where the overlay very slightly appears (see screenshot below). Overlay opacity slider is obviously on max. 
<img width="885" alt="Screenshot 2021-07-02 at 10 58 59" src="https://user-images.githubusercontent.com/47279429/124259772-cb3dd780-db26-11eb-861c-e8bcffb7a0e3.png">
________
### If the value is < 10
I can click on `Find hotspot` and see this temporary visualisation (see screenshot below).

<img width="885" alt="Screenshot 2021-07-02 at 11 04 18" src="https://user-images.githubusercontent.com/47279429/124259867-f0324a80-db26-11eb-88b3-7f3e352a22e7.png">

But then, when clicking `Apply`, it gives me this long error message (open_cv):
```
ERROR: QuPath exception: Could not initialize class org.bytedeco.opencv.global.opencv_imgproc
java.lang.NoClassDefFoundError: Could not initialize class org.bytedeco.opencv.global.opencv_imgproc
    at qupath.opencv.tools.OpenCVTools.findRegionalMaxima(OpenCVTools.java:2276)
    at qupath.lib.analysis.heatmaps.DensityMaps.findHotspots(DensityMaps.java:613)
    at qupath.process.gui.commands.density.DensityMapUI$HotspotFinder.fire(DensityMapUI.java:621)
    at qupath.process.gui.commands.density.DensityMapUI.lambda$createDensityMapAction$2(DensityMapUI.java:324)
    at org.controlsfx.control.action.Action.handle(Action.java:419)
    at org.controlsfx.control.action.Action.handle(Action.java:64)
    at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
    at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:234)
    at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
    at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
    at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
    at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
    at javafx.event.Event.fireEvent(Event.java:198)
    at javafx.scene.Node.fireEvent(Node.java:8889)
    at javafx.scene.control.Button.fire(Button.java:203)
    at com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:208)
    at com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
    at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:247)
    at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
    at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:234)
    at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
    at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
    at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
    at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
    at javafx.event.Event.fireEvent(Event.java:198)
    at javafx.scene.Scene$MouseHandler.process(Scene.java:3856)
    at javafx.scene.Scene.processMouseEvent(Scene.java:1851)
    at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2584)
    at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:409)
    at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:299)
    at java.base/java.security.AccessController.doPrivileged(Unknown Source)
    at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(GlassViewEventHandler.java:447)
    at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:412)
    at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:446)
    at com.sun.glass.ui.View.handleMouseEvent(View.java:556)
    at com.sun.glass.ui.View.notifyMouse(View.java:942)
    at com.sun.glass.ui.mac.MacView.notifyMouse(MacView.java:127)
```
The `Threshold` command works fine and seems to divide the annotation according to the slightly visible overlay, and so does the `Export map` command (so all good).
_________
### If the value is >= 10
Clicking on `Find hotspot`, `Threshold` and `Export map` directly gives me the long open_cv error message copied above.

It might be important to note that when the density radius value is switched back and forth between > and < 10, the density map overlay stops updating sometimes (not sure how to exactly reproduce this though).

**Comment by petebankhead** [comment]
Ah, thanks @melvingelbard  - so it *is* OpenCV-related. That's really good to know, seems to confirm my suspicions although I have no clue why.

It also seems to be specifically the `org.bytedeco.opencv.global.opencv_imgproc` class - assuming object classifiers are working, then `opencv_core` and `opencv_ml` must be ok. If you still have access to the computer, you could check if pixel classifiers work - since they use `opencv_imgproc` as well. If they do, perhaps the problem is more restricted.

If launched from a command line, there might be more detailed errors emitted at some point. On Windows, that can be done simply by using the console exe launcher. On a Mac, it means running something like this from a terminal `/Applications/QuPath.app/Contents/MacOS/QuPath` (exact path might need changed).

**Comment by melvingelbard** [comment]
Yes, the output from my previous post was already taken from the command line, no more info/error was available.
Unfortunately, today I can't seem to be able to reproduce this issue on MacOS, it just miraculously works as expected (??), and so do the pixel classifiers. Launching it again, resetting settings, getting a fresh install, etc.., nothing will get me the original density map exceptions. If someone else experienced the original issue, it would be great to have more details about it.

Worth noticing though is that I sometimes get an Exception when moving the density radius slider left and right (in quite an unpredictable way, therefore not sure how to exactly reproduce this):
```
[ERROR] qupath.lib.gui.dialogs.Dialogs - QuPath exception: Cannot invoke "java.util.List.get(int)" because "minMax" is null
java.lang.NullPointerException: Cannot invoke "java.util.List.get(int)" because "minMax" is null
	at qupath.process.gui.commands.density.DensityMapDialog$ObservableColorModelBuilder.updateDisplayRanges(DensityMapDialog.java:589)
	at qupath.process.gui.commands.density.DensityMapDialog.lambda$new$3(DensityMapDialog.java:201)
	at com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:106)
	at javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:113)
	at javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:147)
	at qupath.process.gui.commands.density.DensityMapDialog$HierarchyClassifierOverlayManager.updateDensityServer(DensityMapDialog.java:884)
	at qupath.process.gui.commands.density.DensityMapDialog$HierarchyClassifierOverlayManager.lambda$updateDensityServer$6(DensityMapDialog.java:863)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by petebankhead** [comment]
Hmmm, for the min/max error do you see similar behavior to the previous problem *after* the exception has been thrown once? It sounds like a bug that can be fixed on the QuPath side with better thread handling. If QuPath recovers and can handle later radius adjustments then I think it's unconnected.

The OpenCV thing is very weird indeed.

**Comment by melvingelbard** [comment]
No, it does look like a separate threading issue in my opinion. I just thought of mentioning it in case someone else experienced the same. I can only see this threading bug happen with the MacBook Pro I have access to, which is the same computer having the original OpenCV issue..

**Comment by petebankhead** [comment]
Reopening this (it was closed automatically).
The macOS problem identified by @melvingelbard should be fixed, but it isn't certain yet whether problems remain elsewhere.

**Comment by petebankhead** [comment]
Closing again, problems reported on Windows now also confirmed fixed in v0.3.0-rc2
